flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorExecutionDeploymentReconciliationTest.java

```diff
@@ -0,0 +1,289 @@
+		return new TestingJobMasterGatewayBuilder()
+			.setRegisterTaskManagerFunction((s, location) -> CompletableFuture.completedFuture(new JMTMRegistrationSuccess(jobManagerResourceId)))
+			.setOfferSlotsFunction((resourceID, slotOffers) -> {
+				slotOfferLatch.trigger();
+				return CompletableFuture.completedFuture(slotOffers);
+			})
+			.setTaskManagerHeartbeatConsumer((resourceID, taskExecutorToJobManagerHeartbeatPayload) -> {
+				ExecutionDeploymentReport executionDeploymentReport = taskExecutorToJobManagerHeartbeatPayload.getExecutionDeploymentReport();
+				deployedExecutionsFuture.add(executionDeploymentReport.getExecutions());
+			})
+			.setUpdateTaskExecutionStateFunction(taskExecutionState -> {
+				if (taskExecutionState.getExecutionState() == ExecutionState.FINISHED) {
+					taskFinishedFuture.complete(null);
+				}
+				return CompletableFuture.completedFuture(Acknowledge.get());
+			})
+			.build();
+	}
+
+	private static TestingResourceManagerGateway setupResourceManagerGateway(CompletableFuture<SlotReport> initialSlotReportFuture) {
+		final TestingResourceManagerGateway testingResourceManagerGateway = new TestingResourceManagerGateway();
+		testingResourceManagerGateway.setSendSlotReportFunction(resourceIDInstanceIDSlotReportTuple3 -> {
+			initialSlotReportFuture.complete(resourceIDInstanceIDSlotReportTuple3.f2);
+			return CompletableFuture.completedFuture(Acknowledge.get());
+		});
+		testingResourceManagerGateway.setRegisterTaskExecutorFunction(input -> CompletableFuture.completedFuture(
+			new TaskExecutorRegistrationSuccess(
+				new InstanceID(),
+				testingResourceManagerGateway.getOwnResourceId(),
+				new ClusterInformation("blobServerHost", 55555))));
+		return testingResourceManagerGateway;
+	}
+
+	private void connectComponentsAndRequestSlot(
+			JobMasterGateway jobMasterGateway,
+			ResourceManagerGateway resourceManagerGateway,
+			TaskExecutorGateway taskExecutorGateway,
+			JobLeaderService jobLeaderService,
+			CompletableFuture<SlotReport> initialSlotReportFuture,
+			AllocationID allocationId) throws Exception {
+		final String jobMasterAddress = "jm";
+		RPC_SERVICE_RESOURCE.getTestingRpcService().registerGateway(jobMasterAddress, jobMasterGateway);
+		RPC_SERVICE_RESOURCE.getTestingRpcService().registerGateway(resourceManagerGateway.getAddress(), resourceManagerGateway);
+
+		// inform the task manager about the job leader
+		jobLeaderService.addJob(jobId, jobMasterAddress);
+		jobManagerLeaderRetriever.notifyListener(jobMasterAddress, UUID.randomUUID());
+		resourceManagerLeaderRetriever.notifyListener(resourceManagerGateway.getAddress(), resourceManagerGateway.getFencingToken().toUUID());
+
+		final Optional<SlotStatus> slotStatusOptional = StreamSupport.stream(initialSlotReportFuture.get().spliterator(), false)
+			.findAny();
+
+		assertTrue(slotStatusOptional.isPresent());
+
+		while (true) {
```

Comment:
Why exactly do we need this while loop here? I think by waiting on `initialSlotReportFuture` it is not needed.